<a href="https://colab.research.google.com/github/nephelim74/machinelearn/blob/main/ml_dz8.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Решающие деревья. Домашнее задание

## Что нужно сделать
Решите задачу классификации оттоковых клиентов телекома: для каждого клиента необходимо определить, перестанет он пользоваться услугами оператора в ближайшее время или нет.

Сделайте первичную обработку данных:

1. Выбросите колонку State.

2. Колонки International plan и Voice mail plan закодируйте следующим образом: Yes замените на 1, а No на 0.

3. Переведите целевую переменную (Churn) в числа: True в 1, а False в 0.

In [14]:
import pandas as pd

# Загружаем данные
url = "https://gbcdn.mrgcdn.ru/uploads/asset/6057170/attachment/326b1a69e5645a4a2f66fdad2177bac7.csv"
data = pd.read_csv(url)
data.head()

State  Account length  Area code International plan Voice mail plan  \
0    KS             128        415                 No             Yes   
1    OH             107        415                 No             Yes   
2    NJ             137        415                 No              No   
3    OH              84        408                Yes              No   
4    OK              75        415                Yes              No   

   Number vmail messages  Total day minutes  Total day calls  \
0                     25              265.1              110   
1                     26              161.6              123   
2                      0              243.4              114   
3                      0              299.4               71   
4                      0              166.7              113   

   Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0             45.07              197.4               99             16.78   
1             27.47              195.5              103             16.62   
2             41.38              121.2              110             10.30   
3             50.90               61.9               88              5.26   
4             28.34              148.3              122             12.61   

   Total night minutes  Total night calls  Total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   Total intl minutes  Total intl calls  Total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   Customer service calls  Churn  
0                       1  False  
1                       1  False  
2                       0  False  
3                       2  False  
4                       3  False

In [15]:

# Удаляем колонку State
data = data.drop("State", axis=1)

# Кодируем колонки International plan и Voice mail plan
data["International plan"] = data["International plan"].map({"Yes": 1, "No": 0})
data["Voice mail plan"] = data["Voice mail plan"].map({"Yes": 1, "No": 0})

# Переводим целевую переменную Churn в числа
data["Churn"] = data["Churn"].astype(int)

# Выводим первые строки для проверки
print(data.head())

   Account length  Area code  International plan  Voice mail plan  \
0             128        415                   0                1   
1             107        415                   0                1   
2             137        415                   0                0   
3              84        408                   1                0   
4              75        415                   1                0   

   Number vmail messages  Total day minutes  Total day calls  \
0                     25              265.1              110   
1                     26              161.6              123   
2                      0              243.4              114   
3                      0              299.4               71   
4                      0              166.7              113   

   Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0             45.07              197.4               99             16.78   
1             27.47              195.5        

Выведите на экран первые строки изменённой таблицы, чтобы проверить, что все преобразования сработали.

In [5]:
data.head()

Account length  Area code  International plan  Voice mail plan  \
0             128        415                   0                1   
1             107        415                   0                1   
2             137        415                   0                0   
3              84        408                   1                0   
4              75        415                   1                0   

   Number vmail messages  Total day minutes  Total day calls  \
0                     25              265.1              110   
1                     26              161.6              123   
2                      0              243.4              114   
3                      0              299.4               71   
4                      0              166.7              113   

   Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0             45.07              197.4               99             16.78   
1             27.47              195.5              103             16.62   
2             41.38              121.2              110             10.30   
3             50.90               61.9               88              5.26   
4             28.34              148.3              122             12.61   

   Total night minutes  Total night calls  Total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   Total intl minutes  Total intl calls  Total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   Customer service calls  Churn  
0                       1      0  
1                       1      0  
2                       0      0  
3                       2      0  
4                       3      0

Создайте матрицу объект-признак X и вектор с целевой переменной (Churn) y.

In [16]:
# ваш код здесь
# Создаем матрицу объект-признак X и вектор целевой переменной y
X = data.drop("Churn", axis=1)
y = data["Churn"]

Проверьте баланс классов. Является ли выборка сбалансированной?

In [17]:
# ваш код здесь
# Проверяем баланс классов
print(y.value_counts())

Churn
0    2850
1     483
Name: count, dtype: int64


Видно, что выборка несбалансированная

Разбейте данные на тренировочную и тестовую части, размер тестовой части — 30%.

In [18]:
from sklearn.model_selection import train_test_split

# ваш код здесь
from sklearn.model_selection import train_test_split

# Разделяем данные на тренировочную и тестовую части (70% - тренировка, 30% - тест)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

Для линейных моделей очень важно масштабирование признаков. Поэтому отмасштабируйте признаки при помощи [StandardScaler](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.StandardScaler.html), обучив метод на тренировочных данных и применив его и к трейну, и к тесту.

In [19]:
from sklearn.preprocessing import StandardScaler

# ваш код здесь
# Масштабируем признаки
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)



После масштабирования признаков X_train и X_test перестали быть объектами типа pd.DataFrame, это неудобно. Приведите их снова к pd.DataFrame.

In [20]:
# ваш код здесь
# Преобразуем обратно в DataFrame для удобства
X_train_scaled = pd.DataFrame(X_train_scaled, columns=X.columns)
X_test_scaled = pd.DataFrame(X_test_scaled, columns=X.columns)

Обучите на тренировочных данных с параметрами по умолчанию:

1) логистическую регрессию;

2) метод опорных векторов (с kernel='linear').

In [21]:
from sklearn.linear_model import LogisticRegression

# Обучаем модель
log_reg = LogisticRegression()
log_reg.fit(X_train_scaled, y_train)

# Предсказание на тесте
y_pred_log_reg = log_reg.predict(X_test_scaled)

In [22]:
from sklearn.svm import SVC

# Обучаем модель
svm = SVC(kernel='linear', probability=True)
svm.fit(X_train_scaled, y_train)

# Предсказание на тесте
y_pred_svm = svm.predict(X_test_scaled)

С помощью обученных моделей сделайте предсказание на тестовых данных.

In [23]:
from sklearn.metrics import roc_auc_score

# ваш код здесь
from sklearn.metrics import roc_auc_score

# Оценка для логистической регрессии
roc_auc_log_reg = roc_auc_score(y_test, log_reg.predict_proba(X_test_scaled)[:, 1])
print("ROC-AUC для логистической регрессии:", roc_auc_log_reg)

# Оценка для SVM
roc_auc_svm = roc_auc_score(y_test, svm.predict_proba(X_test_scaled)[:, 1])
print("ROC-AUC для SVM:", roc_auc_svm)

ROC-AUC для логистической регрессии: 0.8282102961216148
ROC-AUC для SVM: 0.7401816386647192


Теперь обучите решающее дерево с параметрами по умолчанию на тренировочных данных, сделайте предсказание на тесте. Используйте метрику для оценки.

In [24]:
from sklearn.tree import DecisionTreeClassifier

# ваш код здесь
from sklearn.tree import DecisionTreeClassifier

# Обучаем решающее дерево
tree = DecisionTreeClassifier(random_state=42)
tree.fit(X_train, y_train)

# Предсказание на тесте
y_pred_tree = tree.predict(X_test)

# Оценка
roc_auc_tree = roc_auc_score(y_test, tree.predict_proba(X_test)[:, 1])
print("ROC-AUC для решающего дерева:", roc_auc_tree)

ROC-AUC для решающего дерева: 0.8414619219753408


Нужно ли масшабировать признаки для обучения решающего дерева? Проверьте это, заново разбив данные на тренировочную и тестовую части (назовите новые полученные объекты новыми названиями), и проведите обучение, предсказание и оценку метрики до масштабирования и после.

In [26]:
# ваш код здесь
# Разделяем данные заново без масштабирования
X_train_new, X_test_new, y_train_new, y_test_new = train_test_split(X, y, test_size=0.3, random_state=42)

# Обучаем решающее дерево без масштабирования
tree_new = DecisionTreeClassifier(random_state=42)
tree_new.fit(X_train_new, y_train_new)

# Предсказание и оценка
y_pred_tree_new = tree_new.predict(X_test_new)
roc_auc_tree_new = roc_auc_score(y_test_new, tree_new.predict_proba(X_test_new)[:, 1])
print("ROC-AUC для решающего дерева без масштабирования:", roc_auc_tree_new)

ROC-AUC для решающего дерева без масштабирования: 0.8414619219753408


Какая модель справилась с решением задачи?

In [27]:
# ваш ответ здесь
print("ROC-AUC для логистической регрессии:", roc_auc_log_reg)
print("ROC-AUC для SVM:", roc_auc_svm)
print("ROC-AUC для решающего дерева:", roc_auc_tree)
print("ROC-AUC для решающего дерева без масштабирования:", roc_auc_tree_new)

ROC-AUC для логистической регрессии: 0.8282102961216148
ROC-AUC для SVM: 0.7401816386647192
ROC-AUC для решающего дерева: 0.8414619219753408
ROC-AUC для решающего дерева без масштабирования: 0.8414619219753408


### Ответ на вопрос:
Нужно ли масштабировать признаки для решающего дерева?
Нет, решающие деревья не требуют масштабирования, так как они работают с разделением данных на основе порогов, а не расстояний.




### Какая модель справилась лучше?
Финальные результаты:
Логистическая регрессия: ROC-AUC = 0.828

SVM: ROC-AUC = 0.740

Решающее дерево (с масштабированием): ROC-AUC = 0.841

Решающее дерево (без масштабирования): ROC-AUC = 0.841

Анализ результатов:
Решающее дерево показало наилучший результат (ROC-AUC = 0.841):

Результат с масштабированием и без масштабирования одинаковый, что подтверждает, что масштабирование не влияет на качество решающего дерева.

Это ожидаемо, так как деревья работают с порогами, а не с расстояниями.

Логистическая регрессия (ROC-AUC = 0.828):

Результат близок к решающему дереву, но немного хуже.

Логистическая регрессия — это линейная модель, и её результаты могут быть ограничены, если данные нелинейно разделимы.

SVM (ROC-AUC = 0.740):

Самый низкий результат среди всех моделей.

Возможно, линейное ядро (kernel='linear') не подходит для этих данных, или модель требует настройки гиперпараметров.

#### Почему решающее дерево справилось лучше?
Независимость от масштабирования:

Решающие деревья не требуют масштабирования признаков, так как они работают с порогами, а не с расстояниями.

Это делает их более устойчивыми к данным с разными масштабами.

Адаптация к нелинейным зависимостям:

Деревья могут моделировать сложные нелинейные зависимости в данных, что полезно, если связь между признаками и целевой переменной нелинейна.

Работа с категориальными и числовыми признаками:

Деревья легко обрабатывают как числовые, так и категориальные признаки (в вашем случае категориальные признаки уже были закодированы).

